service-worker/worker/src/plugins/dynamic/group.ts

```diff
@@ -0,0 +1,484 @@
+}
+
+/**
+ * Map of strategy names to implementations.
+ */
+export interface DynamicStrategyMap {
+  [strategy: string]: DynamicStrategy;
+}
+
+/**
+ * Represents a specific cache group with a single policy.
+ */
+export class DynamicGroup {
+  /**
+   * A queue of cached URLs, sorted according to the caching configuration (fifo,
+   * lfu, or lru). This is maintained in memory only, and reconstructed by `open`
+   * when loading from saved state.
+   */
+  queue: SortedLinkedList<string>;
+
+  /**
+   * Metadata for entries in this group's cache. Only URLs which exist in `queue`
+   * should have entries in this metadata map.
+   * 
+   * The metadata map is mirrored to a 'metadata' cache entry under this group's
+   * scoped cache, keyed by the request.
+   */
+  metadata: EntryMetadataMap;
+
+  /**
+   * The optimization strategy used for requests in this group. The actual work
+   * of determining whether to used cached responses or continue to the network
+   * is done by the `DynamicStrategy`, not the `DynamicGroup`.
+   */
+  strategy: DynamicStrategy;
+
+  /**
+   * The user-provided manifest which configures the behavior of this dynamic
+   * caching group.
+   */
+  config: GroupManifest;
+
+  /**
+   * Facade which enables unit testing of the cache group.
+   */
+  adapter: NgSwAdapter;
+
+  /**
+   * A cache scoped to this particular dynamic group.
+   */
+  cache: NgSwCache;
+
+  /**
+   * Matchers that will detect URLs which should be handled by this group.
+   */
+  matchers: UrlMatcher[];
+
+  /**
+   * Clock which enables easy unit testing of `DynamicGroup`'s cache expiration
+   * operations through mocking of the current time.
+   */
+  clock: Clock;
+
+  /**
+   * Consumers should use `DynamicGroup.open` instead.
+   */
+  constructor(
+      strategy: DynamicStrategy,
+      config: GroupManifest,
+      adapter: NgSwAdapter,
+      cache: NgSwCache,
+      matchers: UrlMatcher[],
+      metadata: EntryMetadataMap,
+      clock: Clock) {
+    // Obligatory Javaesque assignment of class properties.
+    this.strategy = strategy;
+    this.config = config;
+    this.adapter = adapter;
+    this.cache = cache;
+    this.matchers = matchers;
+    this.metadata = metadata;
+    this.clock = clock;
+
+    // Construct the queue with a comparison strategy based on the expiration
+    // strategy chosen by the user.
+    switch (config.cache.strategy) {
+      case 'fifo':
+        this.queue = new SortedLinkedList<string>(this.fifoCompare.bind(this));
+        break;
+      case 'lfu':
+        this.queue = new SortedLinkedList<string>(this.lfuCompare.bind(this));
+        break;
+      case 'lru':
+        this.queue = new SortedLinkedList<string>(this.lruCompare.bind(this));
+        break;
+      default:
+        throw new Error(`Unknown cache strategy: ${config.cache.strategy}`);
+    }
+    Object.keys(this.metadata).forEach(url => this.queue.insert(url));
```

Comment:
Theoretically, yes, though this is unlikely since the values there were serialized from the same array.

It's okay, though, since even if this happens, the first new insertion will cause a wave of evictions.